src/binding/abi/mpi_abi_util.h

```diff
@@ -0,0 +1,304 @@
+        return MPI_ERRORS_ARE_FATAL;
+    } else if (in == ABI_ERRORS_ABORT) {
+        return MPI_ERRORS_ABORT;
+    } else if (in == ABI_ERRORS_RETURN) {
+        return MPI_ERRORS_RETURN;
+    }
+    return ((MPIR_Errhandler *) in)->handle;
+}
+
+static inline ABI_Errhandler ABI_Errhandler_from_mpi(MPI_Errhandler in)
+{
+    if (in == MPI_ERRHANDLER_NULL) {
+        return ABI_ERRHANDLER_NULL;
+    } else if (in == MPI_ERRORS_ARE_FATAL) {
+        return ABI_ERRORS_ARE_FATAL;
+    } else if (in == MPI_ERRORS_ABORT) {
+        return ABI_ERRORS_ABORT;
+    } else if (in == MPI_ERRORS_RETURN) {
+        return ABI_ERRORS_RETURN;
+    }
+    MPIR_Errhandler *ptr;
+    MPIR_Errhandler_get_ptr(in, ptr);
+    return (ABI_Errhandler) (void *) ptr;
+}
+
+static inline MPI_Group ABI_Group_to_mpi(ABI_Group in)
+{
+    if (ABI_IS_BUILTIN(in)) {
+        if (in == ABI_GROUP_NULL) {
+            return MPI_GROUP_NULL;
+        } else if (in == ABI_GROUP_EMPTY) {
+            return MPI_GROUP_EMPTY;
+        } else {
+            return MPI_GROUP_NULL;
+        }
+    }
+    return ((MPIR_Group *) in)->handle;
+}
+
+static inline ABI_Group ABI_Group_from_mpi(MPI_Group in)
+{
+    if (in == MPI_GROUP_NULL) {
+        return ABI_GROUP_NULL;
+    } else if (in == MPI_GROUP_EMPTY) {
+        return ABI_GROUP_EMPTY;
+    }
+    MPIR_Group *ptr;
+    MPIR_Group_get_ptr(in, ptr);
+    return (ABI_Group) (void *) ptr;
+}
+
+static inline MPI_Info ABI_Info_to_mpi(ABI_Info in)
+{
+    if (in == ABI_INFO_NULL) {
+        return MPI_INFO_NULL;
+    } else if (in == ABI_INFO_ENV) {
+        return MPI_INFO_ENV;
+    }
+    return ((MPIR_Info *) in)->handle;
+}
+
+static inline ABI_Info ABI_Info_from_mpi(MPI_Info in)
+{
+    if (in == MPI_INFO_NULL) {
+        return ABI_INFO_NULL;
+    } else if (in == MPI_INFO_ENV) {
+        return ABI_INFO_ENV;
+    }
+    MPIR_Info *ptr;
+    MPIR_Info_get_ptr(in, ptr);
+    return (ABI_Info) (void *) ptr;
+}
+
+static inline MPI_Message ABI_Message_to_mpi(ABI_Message in)
+{
+    if (in == ABI_MESSAGE_NULL) {
+        return MPI_MESSAGE_NULL;
+    } else if (in == ABI_MESSAGE_NO_PROC) {
+        return MPI_MESSAGE_NO_PROC;
+    }
+    return ((MPIR_Request *) in)->handle;
+}
+
+static inline ABI_Message ABI_Message_from_mpi(MPI_Message in)
+{
+    if (in == MPI_MESSAGE_NULL) {
+        return ABI_MESSAGE_NULL;
+    } else if (in == MPI_MESSAGE_NO_PROC) {
+        return ABI_MESSAGE_NO_PROC;
+    }
+    MPIR_Request *ptr;
+    MPIR_Request_get_ptr(in, ptr);
+    return (ABI_Message) (void *) ptr;
+}
+
+static inline MPI_Op ABI_Op_to_mpi(ABI_Op in)
+{
+    if (ABI_IS_BUILTIN(in)) {
+        /* TODO: error check */
+        return abi_op_builtins[(intptr_t) in & 0xff];
+    }
+    return ((MPIR_Op *) in)->handle;
+}
+
+static inline ABI_Op ABI_Op_from_mpi(MPI_Op in)
+{
+    if (in == MPI_OP_NULL) {
+        return ABI_OP_NULL;
+    }
+    if (HANDLE_IS_BUILTIN(in)) {
+        for (int i = 0; i < ABI_MAX_DATATYPE_BUILTINS; i++) {
```

Comment:
are lines 186 and 187 correct? should they be identical to lines 60 and 61, or substitute datatype for op?

Reply:
Oh, that's because we don't have any test that returns a *builtin-*OP to user.